File: Problem_1.py
# Time Complexity : O(1) for all operations except pop and peek this is amortized O(1)
# Space Complexity : O(N)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No
# we utilize two stack where we are inseting into our delte stack by rmoving in reverse order from our add stack which will populate the delete stack in the expected queue format 

class MyQueue:

    def __init__(self):
        self.addStack = []
        self.deleteStack = []
    def push(self, x: int) -> None:
        self.addStack.append(x)

    def pop(self) -> int:
        self.copyElements()
        return self.deleteStack.pop()

    def peek(self) -> int:
        self.copyElements()
        return self.deleteStack[-1]

    def empty(self) -> bool:
        return not self.addStack and not self.deleteStack
    def copyElements(self):
        if not self.deleteStack:
            while self.addStack:
                self.deleteStack.append(self.addStack.pop())

# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()

File: problem1.go
// Queue using stacks

import (
    "container/list"
)

type Stack struct{
    data *list.List
}

func NewStack() *Stack {
    return &Stack{
        data: list.New(),
    }
}

func (s *Stack) Len() int {
    return s.data.Len()
}

func (s *Stack) Push(x int) {
    s.data.PushBack(x)
}

func (s *Stack) Pop() *list.Element {
    if s.Len() == 0 {
        return nil
    }
    e := s.data.Back()
    s.data.Remove(e)
    return e
}

func (s *Stack) Peek() *list.Element {
    return s.data.Back()
}

type MyQueue struct {
    pushStack *Stack
    popStack *Stack
}

func Constructor() MyQueue {
    return MyQueue {
        pushStack: NewStack(),
        popStack: NewStack(),
    }
}

func (this *MyQueue) Push(x int)  {
    this.pushStack.Push(x)
}


func (this *MyQueue) Pop() int {
    e := this.Peek()
    this.popStack.Pop()
    return e
}


func (this *MyQueue) Peek() int {
    if this.popStack.Len() == 0 {
        for this.pushStack.Len() != 0 {
            this.popStack.Push(this.pushStack.Pop().Value.(int)) 
        }
    }
    e := this.popStack.Peek()
    if e != nil {
        return e.Value.(int)
    }
    return -1
}


func (this *MyQueue) Empty() bool {
    return this.pushStack.Len() == 0 && this.popStack.Len() == 0
}


/**
 * Your MyQueue object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(x);
 * param_2 := obj.Pop();
 * param_3 := obj.Peek();
 * param_4 := obj.Empty();
 */